### ISSUE TITLE
"No projects matched the filters" when dependency filter actually matched

### ISSUE THREAD

**Issue by kasbah** [pnpm/pnpm#5601]
### pnpm version:
7.14.2

### Code to reproduce the issue:

https://github.com/kasbah/pnpm-filter-example

run:

```
pnpm --filter 'one^...' test
```
### Expected behavior:
It just runs the test for `two` and doesn't print a warning. 

### Actual behavior:

It prints a warning:

```
No projects matched the filters ".../pnpm-filter-example" in ".../pnpm-filter-example"
```
But it also (!), runs the test for `two` and exits with a 0 status. It's great that it runs it but I'd prefer it if it didn't print the warning since it seems to have found a project that matched. :slightly_smiling_face: 

### Additional information:

 - `node -v` prints: `v16.18.1`
 - Windows, macOS, or Linux?: Linux

**Comment by kasbah** [comment]
I just noticed that adding a `package.json` in the root works around this issue.

**Comment by anonmily** [comment]
Yep, I had this problem too when using pnpm with nx... Ran the Nx generator for a React library, and it only created a project.json not a package.json in the library. This seems to have caused issues, so just adding in a package.json in the library with the correct package name did the trick.

**Comment by johtso** [comment]
Also having this problem, not sure how I'm supposed to run `pnpm remove` on a specific project with a filter over the dependency names.

**Comment by TomzBench** [comment]
I have an empty package.json {} in my root and then the warning goes away.  Is this neccessary?

**Comment by artemshchirov** [comment]
> Also having this problem, not sure how I'm supposed to run `pnpm remove` on a specific project with a filter over the dependency names.

It looks like the filter uses the packages name (as defined in package.json) rather than its workspace directory.

**Comment by pavi2410** [comment]
For some reason, using `lib` as the filter doesn't work. Is that hard-coded somewhere?
![image](https://github.com/pnpm/pnpm/assets/28837746/bad7e9bb-e4ae-4b43-9a8c-6efb59afdf61)

**Comment by artemshchirov** [comment]
> For some reason, using `lib` as the filter doesn't work. Is that hard-coded somewhere? ![image](https://user-images.githubusercontent.com/28837746/265509966-bad7e9bb-e4ae-4b43-9a8c-6efb59afdf61.png)

"name" field inside package.json of package 

![image](https://github.com/pnpm/pnpm/assets/78075439/a575059c-d6b8-4686-ad35-6a8fb8161803)

**Comment by pavi2410** [comment]
Yes, I got this now. Bit difficult to wrap my head wround.

**Comment by jpenna** [comment]
Shouldn't it use the name defined in `pnpm-workspace.yaml`, matching the folder name?

What are the other use cases for `--filter` that requires it to mathc the package name defined in `package.json`?

**Comment by huevosyitz** [comment]
i also encountered the same. filter uses the 'name' in the package.json to filter. However, you could also filter by directories by using glob filters as stated in the docs here:
https://pnpm.io/filtering#--filter-glob---filter-glob

Also worth noting that you must use double quotes in your glob pattern. single quotes would not work (atleast in my case).

not working:
```
pnpm -F './packages/**' dev
```

working:
```
pnpm -F "./packages/**" dev
```


same goes if you put your pnpm commands in script section in package.json. you need to escape it like so:

```json
{
   "scripts": {
         "dev:packages": "pnpm -F \"./packages/**\" --parallel --stream -r run dev",
  }
}
````

**Comment by eighty4** [comment]
When the filter does not match a workspace project and this message is printed, should the command exit with a non-zero status code? Using `--filter` in a docker build step did not trigger an error even though the build was not occurring.

**Comment by floratmin** [comment]
I encountered the same issue when the depth is only one folder deep.

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: `pnpm --filter "<pkg>^..." run <script>` selects the dependencies and runs, with no "No projects matched the filters" warning.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
